src/OpenFeature/OpenFeatureClient.cs

```diff
@@ -80,76 +85,69 @@ public FeatureClient(IFeatureProvider featureProvider, string name, string versi
                     .ToFlagEvaluationDetails();
 
-                await TriggerAfterHooks(allHooksReversed, hookContext, evaluation, options);
+                await this.TriggerAfterHooks(allHooksReversed, hookContext, evaluation, options);
             }
-            catch (Exception e)
+            catch (FeatureProviderException ex)
             {
-                OpenFeature.Logger.LogError(e, "Error while evaluating flag {FlagKey}", flagKey);
-                // TODO needs to handle error codes being thrown from provider that maps to Errors enums
-                evaluation = new FlagEvaluationDetails<T>(flagKey, defaultValue, e.Message, Reason.Error, string.Empty);
-                await TriggerErrorHooks(allHooksReversed, hookContext, e, options);
+                this._logger.LogError(ex, "Error while evaluating flag {FlagKey}. Error {ErrorType}", flagKey, ex.ErrorTypeDescription);
+                evaluation = new FlagEvaluationDetails<T>(flagKey, defaultValue, ex.ErrorType, Reason.Error, string.Empty);
+                await this.TriggerErrorHooks(allHooksReversed, hookContext, ex, options);
+            }
+            catch (Exception ex)
+            {
+                this._logger.LogError(ex, "Error while evaluating flag {FlagKey}", flagKey);
+                var errorCode = ex is InvalidCastException ? ErrorType.TypeMismatch : ErrorType.General;
+                evaluation = new FlagEvaluationDetails<T>(flagKey, defaultValue, errorCode, Reason.Error, string.Empty);
+                await this.TriggerErrorHooks(allHooksReversed, hookContext, ex, options);
             }
             finally
             {
-                await TriggerFinallyHooks(allHooksReversed, hookContext, options);
+                await this.TriggerFinallyHooks(allHooksReversed, hookContext, options);
             }
 
             return evaluation;
         }
 
-        private static async Task TriggerBeforeHooks<T>(IReadOnlyList<IHook> hooks, HookContext<T> context, FlagEvaluationOptions options)
+        private async Task TriggerBeforeHooks<T>(IReadOnlyList<IHook> hooks, HookContext<T> context, FlagEvaluationOptions options)
         {
             foreach (var hook in hooks)
             {
-                try
+                var resp = await hook.Before(context, options?.HookHints);
+                if (resp != null)
                 {
-                    var resp = await hook.Before(context, options?.HookHints);
-                    if (resp != null)
-                    {
-                        context.EvaluationContext.Merge(resp);
-                    }
-                    else
-                    {
-                        OpenFeature.Logger.LogDebug("Hook {HookName} returned null, nothing to merge back into context", hook.GetType().Name);
-                    }
+                    context.EvaluationContext.Merge(resp);
                 }
-                catch (Exception e)

```

Comment:
So before errors were being swallowed, now they are thrown and handled in the evaluation, which makes sense. :+1: